docs/adrs/0000-composite-actions.md

```diff
@@ -0,0 +1,97 @@
+**Date**: 2021-06-10
+
+**Status**: Accepted
+
+## Context
+
+We released [composite run steps](https://github.com/actions/runner/pull/554) last year which started our journey of reusing steps across different workflow files. To continue that journey, we want to expand composite run steps into composite actions.
+
+We want to support the `uses` steps from workflows in composite actions, including:
+  - Container actions
+  - javascript actions
+  - and even other Composite actions (up to a limit of course!)
+  - The pre and post steps these actions can generate
+
+## Guiding Principles
+
+- Composite Actions should function as a single step or action, no matter how many steps it is composed of or how many levels of recursion it has
+- A workflow author should not need to understand the inner workings of a composite action in order to use it
+- Composite actions should leverage inputs to get values they need, they will not have full access to the `context` objects. The secrets context will **not** be available to composite actions, users will need to pass these values in as an input.
+- Other Actions should **just work** inside a composite action, without any code changes
+
+## Decisions
+
+### Composite Recursion Limit
+
+We will start with supporting a recursion limit of `3`
+That means you can have
+```
+- Composite Action
+  - Nested Composite Action
+    - Deep Nested Composite Action
+      - uses action step
+      - runs action step
+```
+- We are free to bump this limit in the future, the code will be written to just require updating a variable. If the graph evaluates beyond the recursion limit, the job will fail in the pre-job phase (The `Set up job` step).
+- A composite actions interface is its inputs and outputs, nothing else is carried over when invoking recursively.
+
+### Pre/Post Steps in nested Actions
+
+- We do not plan on adding the ability to configure a customizable pre or post step for composite actions at this time. However, we will execute the pre and post steps of any actions referenced in a composite action.
+- Composite actions will generate a single pre-step and post-step for the entire composite action, even if there are multiple pre-steps and post-steps in the referenced actions.
```

Comment:
we need to be careful to make sure intra-state doesnt get clobbered (one logical step in the UI, but multiple steps wrt intra-state)

Reply:
ok, reading further below it looks like you address this already in the "set-state" section